conformance-test-framework/src/main/java/org/jspecify/conformance/ExpectedFact.java

```diff
@@ -0,0 +1,167 @@
+          Pattern.compile("test:sink-type:[^:]+:.*"));
+
+  private final String fact;
+  private final long factLineNumber;
+
+  ExpectedFact(Path file, long lineNumber, String fact, long factLineNumber) {
+    super(file, lineNumber);
+    this.fact = fact;
+    this.factLineNumber = factLineNumber;
+  }
+
+  /**
+   * Returns an expected fact representing that the source type cannot be converted to the sink type
+   * in any world.
+   */
+  public static String cannotConvert(String sourceType, String sinkType) {
+    return String.format("test:cannot-convert:%s to %s", sourceType, sinkType);
+  }
+
+  /** Returns an expected fact representing an expected expression type. */
+  public static String expressionType(String expressionType, String expression) {
+    return String.format("test:expression-type:%s:%s", expressionType, expression);
+  }
+
+  /** Returns an expected fact representing that an annotation is not relevant. */
+  public static String irrelevantAnnotation(String annotationType) {
+    return String.format("test:irrelevant-annotation:%s", annotationType);
+  }
+
+  /** Returns an expected fact representing that an annotation is not relevant. */
+  public static String sinkType(String sinkType, String sink) {
+    return String.format("test:sink-type:%s:%s", sinkType, sink);
+  }
+
+  /**
+   * Returns {@code true} if {@code fact} is a legacy {@code jspecify_nullness_mismatch} assertion.
+   */
+  public boolean isNullnessMismatch() {
+    return NULLNESS_MISMATCH.matcher(getFactText()).matches();
+  }
+
+  @Override
+  String getFactText() {
+    return fact;
+  }
+
+  /** Returns the line number in the input file where the expected fact is. */
+  long getFactLineNumber() {
+    return factLineNumber;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof ExpectedFact)) {
+      return false;
+    }
+    ExpectedFact that = (ExpectedFact) o;
+    return this.getFile().equals(that.getFile())
+        && this.getLineNumber() == that.getLineNumber()
+        && this.fact.equals(that.fact)
+        && this.factLineNumber == that.factLineNumber;
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(getFile(), getLineNumber(), fact);
+  }
+
+  @Override
+  public String toString() {
+    return toStringHelper(this)
+        .add("file", getFile())
+        .add("lineNumber", getLineNumber())
+        .add("fact", fact)
+        .add("factLineNumber", factLineNumber)
+        .toString();
+  }
+
+  /** Reads {@link ExpectedFact}s from comments in a file. */
+  static class Reader {
+
+    private static final Pattern EXPECTATION_COMMENT =
+        Pattern.compile(
+            "\\s*// "
+                + ("(?<expectation>"
+                    + ASSERTION_PATTERNS.stream().map(Pattern::pattern).collect(joining("|"))
+                    + ")")
+                + "\\s*");
+
+    private final Map<Long, String> facts = new HashMap<>();
```

Comment:
These two are fields to avoid the allocation of a new `HashMap` for multiple calls of `readExpectedFacts` on the same `Reader`?

Reply:
Sort of. they don't need to be fields yet, but it'll help when I add support for test names. I'll remove them from this.